WIP: Version 3.1.0

tferr opened the pull request:
This will implement several features to modernize SNT
- Productivity tools for efficient tracing:
  - Snap-to window cursor (should address  #1)
  - Extended keyboard shortcuts and fixes for #15
  - Auto-canvas activation, and arrangement of windows that is aware of multiple displays
- Color schemes for path rendering
- Cleanup of GUI and code.

Will merge it ASA as I find more time to test things up. Any feedback more than welcomed.

tferr committed:
Allow paths to have ad-hoc colors

tferr committed:
Render ad-hoc colors in 3D viewer

tferr committed:
Implement GUI for new coloring options

tferr committed:
Make TracerCanvas and plugin aware of custom coloring schemes

tferr committed:
Update Export command

tferr committed:
Implement option in GUI to revert to legacy behavior

tferr committed:
Revamp GUI:
- Organize it in sections
- Make it more interactive
- Improve resizing behavior

tferr committed:
ROI export: Prefer default path name

tferr committed:
Improve startup behavior: Do not impose LUT and collect options in a single prompt

tferr committed:
Improve startup: Arrange windows at startup

tferr committed:
Code cleanup: Organize Imports

tferr committed:
Code Cleanup: Apply http://imagej.net/Coding_style

tferr committed:
Fix #15

tferr committed:
Organize GUI: Aggregate tracking-related commands in a dedicated menu

tferr committed:
Implement auto-canvas activation as soon as cursor enters it

tferr committed:
Get rid of those unrelated file icons in JTree

tferr committed:
Make image ready for tracing at startup

tferr committed:
Implement method for local snap-to-cursor
As is it is only aware of local maxima. In the future, we should improve
it to be aware of e.g., tubbiness image

tferr committed:
Add snap-to feature to GUI

tferr committed:
Allow toggling of snap-to cursor via Keyboard

tferr committed:
Prune dead code

tferr committed:
GUI Tweaks: make shortcut more visible, correct alignments..

tferr committed:
Correct outdated instructions in error messages.
While at it make everything tidier.

tferr committed:
Implement shortcut for canceling path

tferr committed:
Use a dedicated class to manage preferences

tferr committed:
Implement the Tracing>Options... command

tferr committed:
Update methods for loading of Prefs before displaying main GUI

tferr committed:
FIX typo, inverted prefs and such

tferr committed:
Handle all logging and errors consistently
This is all done through bulk search and replace

tferr committed:
Move static method to Utilities class

tferr committed:
Cleanup Fill window
(Which surprisingly enough was not functional at all, despite the absence of complaints)

tferr committed:
Allow tracing on snap-to-cursor position

tferr committed:
Prune legacy code

tferr committed:
Minor cleanup

tferr committed:
Extend color schemes to fitted paths

tferr committed:
Prefer package logger

imagejan commented on src/main/java/tracing/SNT.java:
```diff
@@ -0,0 +1,26 @@
+	}
+
+	protected static void log(final String string) {
+		System.out.println("[SNT] " + string);
```
Why not use a [SciJava `LogService`](http://javadoc.imagej.net/SciJava/index.html?org/scijava/log/LogService.html) for error and logging messages? I see that you're depending on SciJava already anyway.

tferr replied:
@imagejan, @ctrueden: Yeah, that was the idea behind this code cleanup: Move all the hundreds of ```System.out.println``` calls in Simple Neurite Tracer to a single utility class so that things can be more more customizable. I started using [LogService in BAR](https://github.com/tferr/Scripts/blob/master/BAR/src/main/java/bar/Runner.java#L58-L80) while ago, and thought about it here but the utility [SNT class](https://github.com/tferr/Simple_Neurite_Tracer/blob/dev/src/main/java/tracing/SNT.java)  just contains static methods and Simple Neurite Tracer is a mammoth of a plugin stuck in IJ1. The only way I thought about implementing it (after e.g., this [forum thread](http://forum.imagej.net/t/gui-stalling-and-eventually-freezing-while-trakem2-generates-mipmaps/?u=tferr)), was using something like:

```java
@Plugin(type = ContextCommand.class)
public class SNT {

	@Parameter
	private static Context context;

	@Parameter
	private static LogService logService;

	private SNT() {
		// private constructor to prevent instantiation of class
	}
	
	private static void initLog() {
		if (context == null)
			context = (Context) IJ.runPlugIn("org.scijava.Context", "");
		if (logService == null)
			logService = context.getService(LogService.class);
	}

	protected static void log(final String string) {
		if (logService == null)
			initLog();
		logService.debug("[SNT] " + string);
	}

}
```
Would you guys be OK with something like the above? Is this even good practice?

ctrueden replied:
@tferr Actually, I think that looks good. It's precisely how you have to access the SciJava context from IJ1-land. There is no need for the `@Parameter` annotations though; they don't really do anything here. The only other consideration is whether the SNT class needs to be thread-safe. If not, what you have will work. But if you need thread safety, you can use a double-checked locking pattern like many of the SciJava services do for their lazy initialization. (You can do `git grep 'lazy init'` to find lots of examples in the scijava-common codebase.)

tferr replied:
OK. Thanks a lot for the input. It was done on e4528a9. I've also started to replace `IJ.log()` calls throughout the several classes of the plugin, and initiated descriminating the type of debug level (error, info, warning), etc. The SNT class now looks [like this](https://github.com/tferr/Simple_Neurite_Tracer/blob/dev/src/main/java/tracing/SNT.java).

Note that there are _lots_ of things that could be moved to scijava right now (namely routines dealing with files and routines handling the coloring of paths) that should adopt [scijava.util](http://javadoc.imagej.net/SciJava/index.html?org/scijava/util/). It would be fairly easy, but IMHO that effort should be initiated once SNT is rewritten for ImageJ2.

tferr committed:
More GUI cleanup: Organize menus, prompts, errors and logs

tferr committed:
Allow auto-loading of existing traces

tferr committed:
Add option for compressed/uncompressed XML output

tferr committed:
Adjust default preferences to something more sensible

tferr committed:
Store custom path colors in traces files

tferr committed:
ensure the same display range in xz zy views
This should fix the nasty rendering of 16-bit images  on these panes

tferr committed:
Use LogService in debug mode

tferr committed:
Use Escape as a shortcut for "Abandon Search"

tferr committed:
Save preferences in exitRequested()

tferr committed:
Allow Filling operations only if image is available

tferr committed:
Allow visibility of 3D viewer to be toggled

tferr committed:
Use snap diameters in prompt not radius

tferr committed:
Ensure snap window dimensions are properly accessed across threads

tferr committed:
Make the "Arrange windows" command aware of multiple displays (or at least try)

tferr committed:
Monitor closure of 3D Viewer

tferr committed:
Make menu entries uniform

tferr committed:
Use SNT for IJ log calls

tferr committed:
Highlight paths with custom colors when selected in 3D viewer

tferr committed:
Ensure choice for 3D viewer is respected at startup

tferr committed:
Code Cleanup: Apply http://imagej.net/Coding_style

tferr committed:
Ensure prompt is visible when using autoCanvasActivation

tferr committed:
Specify spacing units in SWC file header, and other useful "metadata.
While at it fix duplicated code.

tferr committed:
Allow customized lists of paths to be rendered as topographic skeleton

tferr committed:
Avoid NPE at startup

tferr committed:
makePathVolume method aware of subsets of paths

tferr committed:
Keep a reference of original image title in skeleton stack

tferr committed:
Perform all skeleton analysis through a dedicated plugin:
Major advantages of this implementation: ROI/SWC filtering and summary statistics

tferr committed:
Store SWC type in CSV file

tferr committed:
Do our best to avoid the propagation of SWC in pixel coordinates

tferr committed:
GUI access to new plugin

tferr committed:
Simplify (a lot!) method transferring Path-ROIs to Overlay

tferr committed:
Revamp command exporting Path-ROIs

tferr committed:
Sort menu entries, More reasonable GUI entries, Update imports and minor cleanups

tferr committed:
Add URL for Tubuluar Geodesics

tferr committed:
Skip "empty" runs in a couple of menu commands
